File: C++/judgeSquareSum.cpp
/**
 * @File    : judgeSquareSum.cpp
 * @Brief   : 平方数之和
 * @Link    : https://leetcode-cn.com/problems/sum-of-square-numbers/
 * @Author  : Wei Li
 * @Date    : 2021-07-07
*/

#include <iostream>
#include <cmath>
#include <memory>

class Solution
{
public:
    bool judgeSquareSum(int c)
    {
        long left = 0;
        long right = std::sqrt(c);
        while (left <= right)
        {
            long sum = left * left + right * right;
            if (sum == c)
            {
                return true;
            }
            else if (sum > c)
            {
                right--;
            }
            else
            {
                left++;
            }
        }
        return false;
    }
};


// -----------------------------
int main(int argc, char** argv)
{
    // int c = 5;
    // int c = 4;
    int c = 3;
    // int c = 2;
    // int c = 1;

    // auto solution = Solution();
    // Solution solution;
    auto solution = std::make_unique<Solution>();
    bool check = solution->judgeSquareSum(c);
    // bool check = solution.judgeSquareSum(c);
    std::cout << "The solution of this problem is : " << std::boolalpha << check << std::endl;
    
    return 0;
}

File: Python/judgeSquareSum.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: judgeSquareSum.py
@Function: 平方数之和
@Link: https://leetcode-cn.com/problems/sum-of-square-numbers/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-07
"""

import math

class Solution:
    def judgeSquareSum(self, c):
        if c == 0:
            return True

        # ---------------------------------
        # for a in range(1, int(math.sqrt(c) + 1)):
        #     b = c - a**2
        #     if int(math.sqrt(b))**2 == b:
        #         return True
        # ---------------------------------

        # ---------------------------------
        # 双指针策略
        # ---------------------------------
        left, right = 0, int(math.sqrt(c))
        while left <= right:
            sum = left**2 + right**2
            if sum == c:
                return True
            elif sum > c:
                right -= 1
            else:
                left += 1
        # ---------------------------------
        return False


# -------------------------
if __name__ == "__main__":
    c = 5

    solution = Solution()
    check = solution.judgeSquareSum(c)
    print(check)